File: dp/knapsack_problem/cpp/knapsack.cpp
#include <iostream>

using namespace std;

// Inputs taken by the function:

// Values (stored in array v)
// Weights (stored in array w)
// Number of distinct items (n)
// Knapsack capacity (W)

int fill_knapsack(int v[], int w[], int n, int W)
{
	// T[i][j] stores the maximum value of knapsack having weight less than equal to j with only first i items considered.
    int T[n+1][W+1];
 
	for (int j = 0; j <= W; j++)
		T[0][j] = 0;                           //Since no. of items is zero so value of knapsack will be zero (Base Case)
	
	for (int i = 1; i <= n; i++)
	{
		// We consider all weights from 0 to maximum capacity W
		for (int j = 0; j <= W; j++)
		{
			// We don't include ith element if j-w[i-1] is negative
			if (w[i-1] > j)
				T[i][j] = T[i-1][j];
			else
			// We find maximum value we get by excluding or including the ith item
				T[i][j] = max(T[i-1][j],                         //Excluding the item
							T[i-1][j-w[i-1]] + v[i-1]);          //Including the item
		}
	}
	
	// We return the maximum value
    return T[n][W];
}
 
int main()
{
	// Input: set of items each with a weight and a value
	int v[] = { 20, 5, 10, 40, 15, 25};
	int w[] = {  1, 2,  3,  8,  7, 4};
	
	// Knapsack capacity
	int W = 10;

	// Number of items
	int n = sizeof(v) / sizeof(v[0]);

	cout << "The value of the knapsack is " << fill_knapsack(v, w, n, W) << endl;

    return 0;
}

File: dp/knapsack_problem/java/Knapsack.java
public class Knapsack {

	public static void main(String[] args) {
		// Initialize the algorithm
		new Knapsack().begin();
	}

	void begin() {
		int values[] = { 20, 5, 10, 40, 15, 25 };
		int weights[] = { 1, 2, 3, 8, 7, 4 };

		int capacity = 10;
		int n = values.length;

		System.out.println(knapSack(values, weights, n, capacity));
	}

	/** 0/1 Knapsack method
	*@param values - Values stored in array
	*@param weights - Weights stored in array
	*@param n - number of distinct items
	*@param capacity - Knapsack capacity
	*@return max value
	*/

	int knapSack(int values[], int weights[], int n, int capacity) {

		// create a new matrix Items are row and weight are column
		int K[][] = new int[n + 1][capacity + 1];

		//initialize tha values of bag items and capacity
		for (int i = 0; i < n; i++) {
			K[i][0] = 0;
    }

    for (int w = 0; w < capacity; w++) {
			K[0][w] = 0;
    }

		for (int i = 1; i <= n; i++) {
			for (int w = 1; w <= capacity; w++) {
				if (weights[i - 1] <= w) {
					//Find maximum value get by excluding or including the item
					K[i][w] = Math.max(
							values[i - 1] + K[i - 1][w - weights[i - 1]],  //including the item
							K[i - 1][w] // excluding the item
			                             ); 
				} else {
					//If current item weight is more than the running weight, just keep going without the current item
					K[i][w] = K[i - 1][w];
				}
			}
		}
		//return the maximum value
		return K[n][capacity];
	}
}
